packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -232,10 +233,139 @@ export default createRule<Options, MessageId>({
+        if (
+          !(
+            returnTypeAnnotation != null &&
+            ts.isTypePredicateNode(returnTypeAnnotation) &&
+            // This eliminates things like `x is string` and `asserts x is T`
+            // leaving us with just the `asserts x` cases.
+            returnTypeAnnotation.type == null &&
+            // I think this is redundant but, still, it needs to be true
+            returnTypeAnnotation.assertsModifier != null
+          )
+        ) {
+          return undefined;
+        }
+
+        const assertionTarget = returnTypeAnnotation.parameterName;
+        if (assertionTarget.kind !== ts.SyntaxKind.Identifier) {
+          // This can happen when asserting on `this`. Ignore!
+          return undefined;
+        }
+
+        // If the first parameter is `this`, skip it, so that our index matches
+        // the index of the argument at the call site.
+        const firstParameter = declaration.parameters.at(0);
+        const nonThisParameters =
+          firstParameter?.name.kind === ts.SyntaxKind.Identifier &&
+          firstParameter.name.text === 'this'
+            ? declaration.parameters.slice(1)
+            : declaration.parameters;
+
+        // Don't bother inspecting parameters past the number of
+        // arguments we have at the call site.
+        const checkableNonThisParameters = nonThisParameters.slice(
+          0,
+          checkableArguments.length,
+        );
+
+        let assertedParameterIndexForThisSignature: number | undefined;
+        for (const [index, parameter] of checkableNonThisParameters.entries()) {
+          if (parameter.dotDotDotToken != null) {
+            // Cannot assert a rest parameter, and can't have a rest parameter
+            // before the asserted parameter. It's not only a TS error, it's
+            // not something we can logically make sense of, so give up here.
+            return undefined;
+          }
+
+          if (parameter.name.kind !== ts.SyntaxKind.Identifier) {
+            // Only identifiers are valid for assertion targets, so skip over
+            // anything like `{ destructuring: parameter }: T`
+            continue;
+          }
+
+          // we've found a match between the "target"s in
+          // `function asserts(target: T): asserts target;`
+          if (parameter.name.text === assertionTarget.text) {
+            assertedParameterIndexForThisSignature = index;
+            break;
+          }
+        }
+
+        if (assertedParameterIndexForThisSignature == null) {
+          // Didn't find an assertion target in this signature that could match
+          // the call site.
+          return undefined;
+        }
+
+        if (
+          assertedParameterIndex != null &&
+          assertedParameterIndex !== assertedParameterIndexForThisSignature
+        ) {
+          // The asserted parameter we found for this signature didn't match
+          // previous signatures.
+          return undefined;
+        }
```

Comment:
[Refactor] Nit: the if statement below contains `assertedParameterIndex != null && assertedParameterIndex !== assertedParameterIndexForThisSignature`. This means that `assertedParameterIndexForThisSignature` is non-nullable, since  we require it to be equal to the numeric `assertedParameterIndex`.

The first if statements seems redundant:

```suggestion
        if (
          assertedParameterIndex != null &&
          assertedParameterIndex !== assertedParameterIndexForThisSignature
        ) {
          // The asserted parameter we found for this signature didn't match
          // previous signatures.
          return undefined;
        }
```

Reply:
Hmmmmm. You're right that the runtime behavior is the same with or without the branch. However, it's a conceptually different reason for terminating, and, therefore, preserving it ensures code coverage for the different possibilities.

I think I have a slight preference for it to stay the way it is, but I will change if you prefer for the checks to be combined. 

*** 

Another example of something related is here:
https://github.com/typescript-eslint/typescript-eslint/blob/6d680202c7495359561f31273bd4e33e65c17324/packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts#L106-L122
While the branches are identical, the are identical for very different reasons (in the first case, we want the `T` in `Array<T>`, which coincidentally is the 0th type argument... note - it would probably be better written as `firstParamType.getNumberIndexType()`), whereas, in the case of the tuple type, we want the type specifically at the 0th position, i.e. `[string, number]`.